addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/connector/RFXComStreamReader.java

```diff
@@ -37,89 +36,50 @@ public RFXComStreamReader(RFXComBaseConnector connector, InputStream in) {
-        }
-    }
-
-    @Override
     public void run() {
-        final int dataBufferMaxLen = Byte.MAX_VALUE;
-
-        byte[] dataBuffer = new byte[dataBufferMaxLen];
-
-        int msgLen = 0;
-        int index = 0;
-        boolean startFound = false;
-
         logger.debug("Data listener started");
 
-        try {
-
-            byte[] tmpData = new byte[20];
-            int len = -1;
-
-            while (!interrupted) {
-
-                if ((len = in.read(tmpData)) > 0) {
-
-                    byte[] logData = Arrays.copyOf(tmpData, len);
-                    logger.trace("Received data (len={}): {}", len, DatatypeConverter.printHexBinary(logData));
-
-                    for (int i = 0; i < len; i++) {
-
-                        if (index > dataBufferMaxLen) {
-                            // too many bytes received, try to find new
-                            // start
-                            startFound = false;
-                        }
-
-                        if (!startFound && tmpData[i] > 0) {
-
-                            startFound = true;
-                            index = 0;
-                            dataBuffer[index++] = tmpData[i];
-                            msgLen = tmpData[i] + 1;
-
-                        } else if (startFound) {
-
-                            dataBuffer[index++] = tmpData[i];
-
-                            if (index == msgLen) {
-                                sendMessage(dataBuffer, msgLen);
-
-                                // find new start
-                                startFound = false;
-                            }
+        final int MAX_READ_TIMEOUTS = 4;
+        byte[] buf = new byte[Byte.MAX_VALUE];
+
+        // The stream has (or SHOULD have) a read timeout set. Taking a
+        // read timeout (read returns 0) between packets gives us a chance
+        // to check if we've  been interrupted. Read interrupts during a
+        // packet are ignored but if too many occur we take it as meaning
+        // the RFXCOM has become missing presumed dead.
+        while (!Thread.interrupted()) {
+            try {
+                // First byte tells us how long the packet is
+                int bytesRead = in.read(buf, 0, 1);
+                int packetLength = buf[0];
+
+                if (bytesRead > 0 && packetLength > 0) {
+                    // Now read the rest of the packet
+                    int bufferIndex = 1;
+                    int readTimeoutCount = 1;
+
+                    while (bufferIndex <= packetLength) {
+                        int bytesRemaining = packetLength - bufferIndex + 1;
+                        bytesRead = in.read(buf, bufferIndex, bytesRemaining);
+
+                        if (bytesRead > 0) {
+                            bufferIndex += bytesRead;
+                        } else if (readTimeoutCount++ == MAX_READ_TIMEOUTS) {
+                            throw new IOException("Timeout during packet read");
```

Comment:
> A continue would be fatal.
Correct, I missed the second nested loop. So what we actually need to do to avoid throwing and catching an exception in the same method is breaking the inner loop and continue-the outer loop. Java has ways for that but those are a little uncommon. 
_The most proper way is imho using more methods, I think we learned programming from different persons because I always like more methods and like them between 2 & 10 lines._

Thanks for the complete explanation of the error handling 👍